Network pool creation with zero VLAN for VMotion and VSAN pools fails

### Code of Conduct

- [X] I have read and agree to the [Code of Conduct](https://github.com/vmware/terraform-provider-vcf/blob/main/CODE_OF_CONDUCT.md).
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [ ] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Terraform

v1.7.3

### Terraform Provider

v0.9.0

### VMware Cloud Foundation

5.2

### Description

This bug is similar to https://github.com/vmware/terraform-provider-vcf/issues/128.

Creating Network Pools with zero VLAN ID for  vMOTION and vSAN  is not allowed because the provider passes a NULL value to SDDC Network Pool creation API.

The fix will be same as the fix for https://github.com/vmware/terraform-provider-vcf/issues/128 where in it was not allowing zero vlan for GeneveID during bringup.
The fix will be in SDK

### Affected Resources or Data Sources

network_pool


### Terraform Configuration

resource "vcf_network_pool" "eng_pool" {
  name = "engineering-pool"
  network {
    gateway = "10.0.4.253"
    mask    = "255.255.255.0"
    mtu     = 8940
    subnet  = "10.0.4.0"
    type    = "VSAN"
    vlan_id = 0
    ip_pools {
      start = "10.0.4.51"
      end   = "10.0.4.81"
    }
  }
   network {
    gateway = "10.0.8.253"
    mask    = "255.255.255.0"
    mtu     = 8940
    subnet  = "10.0.8.0"
    type    = "vMotion"
    vlan_id = 0
    ip_pools {
      start = "10.0.8.51"
      end   = "10.0.8.81"
    }
  }
}


### Debug Output


2024-05-28T12:12:06.330-0700 [ERROR] provider.terraform-provider-vcf_v0.9.0: Response contains error diagnostic: tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/diag/diagnostics.go:58 diagnostic_severity=ERROR tf_resource_type=vcf_network_pool diagnostic_detail="" tf_req_id=fd84f6c2-4b57-ab5e-28c4-b0bfa318a276 @module=sdk.proto diagnostic_summary="[POST /v1/network-pools][400] createNetworkPoolBadRequest  &{Arguments:[] Causes:[] Context:map[] ErrorCode:REST_INVALID_API_INPUT ErrorType: Message:Invalid input NestedErrors:[0xc000464e60] ReferenceToken:R345HG RemediationMessage:}" tf_provider_addr=provider tf_proto_version=5.6 timestamp=2024-05-28T12:12:06.330-0700
2024-05-28T12:12:06.331-0700 [ERROR] vertex "vcf_network_pool.eng_pool" error: [POST /v1/network-pools][400] createNetworkPoolBadRequest  &{Arguments:[] Causes:[] Context:map[] ErrorCode:REST_INVALID_API_INPUT ErrorType: Message:Invalid input NestedErrors:[0xc000464e60] ReferenceToken:R345HG RemediationMessage:}
╷
│ Error: [POST /v1/network-pools][400] createNetworkPoolBadRequest  &{Arguments:[] Causes:[] Context:map[] ErrorCode:REST_INVALID_API_INPUT ErrorType: Message:Invalid input NestedErrors:[0xc000464e60] ReferenceToken:R345HG RemediationMessage:}


### Panic Output

_No response_

### Expected Behavior

Successful network pool creation

### Actual Behavior

Network pool creation fails

### Steps to Reproduce

Fails to create Network pool


### Environment Details

_No response_

### Screenshots

_No response_

### References

_No response_

--- a/docs/resources/edge_cluster.md
+++ b/docs/resources/edge_cluster.md
@@ -36,21 +36,21 @@ Review the documentation for VMware Cloud Foundation for more information about 
 - `audit_password` (String) Audit user password for the NSX manager
 - `edge_node` (Block List, Min: 1) The nodes in the edge cluster (see [below for nested schema](#nestedblock--edge_node))
 - `form_factor` (String) One among: XLARGE, LARGE, MEDIUM, SMALL
-- `high_availability` (String) One among: ACTIVE_ACTIVE, ACTIVE_STANDBY
 - `mtu` (Number) Maximum transmission unit size for the cluster
 - `name` (String) The name of the edge cluster
 - `profile_type` (String) One among: DEFAULT, CUSTOM. If set to CUSTOM a 'profile' must be provided
 - `root_password` (String) Root user password for the NSX manager
-- `routing_type` (String) One among: EBGP, STATIC
-- `tier0_name` (String) Name for the Tier-0 gateway
-- `tier1_name` (String) Name for the Tier-1 gateway
 
 ### Optional
 
 - `asn` (Number) ASN for the cluster
+- `high_availability` (String) One among: ACTIVE_ACTIVE, ACTIVE_STANDBY
 - `internal_transit_subnets` (List of String) Subnet addresses in CIDR notation that are used to assign addresses to logical links connecting service routers and distributed routers
 - `profile` (Block List, Max: 1) The specification for the edge cluster profile (see [below for nested schema](#nestedblock--profile))
+- `routing_type` (String) One among: EBGP, STATIC
 - `skip_tep_routability_check` (Boolean) Set to true to bypass normal ICMP-based check of Edge TEP / host TEP routability (default is false, meaning do check)
+- `tier0_name` (String) Name for the Tier-0 gateway
+- `tier1_name` (String) Name for the Tier-1 gateway
 - `tier1_unhosted` (Boolean) Select whether Tier-1 being created per this spec is hosted on the new Edge cluster or not (default value is false, meaning hosted)
 - `timeouts` (Block, Optional) (see [below for nested schema](#nestedblock--timeouts))
 - `transit_subnets` (List of String) Transit subnet addresses in CIDR notation that are used to assign addresses to logical links connecting Tier-0 and Tier-1s
@@ -80,8 +80,18 @@ Required:
 Optional:
 
 - `first_nsx_vds_uplink` (String) The name of the first NSX-enabled VDS uplink
+- `management_network` (Block List, Max: 1) The management network which will be created for this node (see [below for nested schema](#nestedblock--edge_node--management_network))
 - `second_nsx_vds_uplink` (String) The name of the second NSX-enabled VDS uplink
 - `uplink` (Block List) Specifications of Tier-0 uplinks for the edge node (see [below for nested schema](#nestedblock--edge_node--uplink))
+
+<a id="nestedblock--edge_node--management_network"></a>
+### Nested Schema for `edge_node.management_network`
+
+Required:
+
+- `portgroup_name` (String) The name of the portgroup
+- `vlan_id` (Number) The VLAN ID for the portgroup
+
 
 <a id="nestedblock--edge_node--uplink"></a>
 ### Nested Schema for `edge_node.uplink`
@@ -102,7 +112,7 @@ Required:
 
 - `asn` (Number) ASN
 - `ip` (String) IP address
-- `password` (String) Password - can be empty.
+- `password` (String) Password
 
 
 

--- a/docs/resources/network_pool.md
+++ b/docs/resources/network_pool.md
@@ -44,6 +44,10 @@ The following data is prerequisite for creating a new Network Pool
 <a id="nestedblock--network"></a>
 ### Nested Schema for `network`
 
+Required:
+
+- `vlan_id` (Number) VLAN ID associated with the network
+
 Optional:
 
 - `gateway` (String) Gateway for the network
@@ -52,7 +56,6 @@ Optional:
 - `mtu` (Number) Gateway for the network
 - `subnet` (String) Subnet associated with the network
 - `type` (String) Network Type of the network
-- `vlan_id` (Number) VLAN ID associated with the network
 
 <a id="nestedblock--network--ip_pools"></a>
 ### Nested Schema for `network.ip_pools`
@@ -70,5 +73,3 @@ Optional:
 Optional:
 
 - `create` (String)
-
-

--- a/go.mod
+++ b/go.mod
@@ -9,7 +9,7 @@ require (
 	github.com/hashicorp/terraform-plugin-log v0.9.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.34.0
 	github.com/stretchr/testify v1.9.0
-	github.com/vmware/vcf-sdk-go v0.3.1
+	github.com/vmware/vcf-sdk-go v0.3.2
 	golang.org/x/exp v0.0.0-20230626212559-97b1e661b5df
 )
 

--- a/internal/nsx_edge_cluster/edge_cluster_operations.go
+++ b/internal/nsx_edge_cluster/edge_cluster_operations.go
@@ -30,6 +30,7 @@ func GetNsxEdgeClusterCreationSpec(data *schema.ResourceData) *models.EdgeCluste
 	mtu := int32(data.Get("mtu").(int))
 	asn := int64(data.Get("asn").(int))
 	tier1Unhosted := data.Get("tier1_unhosted").(bool)
+	skipTepRoutabilityCheck := data.Get("skip_tep_routability_check").(bool)
 
 	transitSubnets := resource_utils.ToStringSlice(data.Get("transit_subnets").([]interface{}))
 	internalTransitSubnets := resource_utils.ToStringSlice(data.Get("internal_transit_subnets").([]interface{}))
@@ -62,6 +63,7 @@ func GetNsxEdgeClusterCreationSpec(data *schema.ResourceData) *models.EdgeCluste
 		Tier1Name:                     tier1Name,
 		Tier1Unhosted:                 tier1Unhosted,
 		TransitSubnets:                transitSubnets,
+		SkipTepRoutabilityCheck:       skipTepRoutabilityCheck,
 	}
 
 	return spec
@@ -162,6 +164,16 @@ func getNodeSpec(node map[string]interface{}) *models.NsxTEdgeNodeSpec {
 		SecondNsxVdsUplink: secondVdsUplink,
 		InterRackCluster:   &interRackCluster,
 		UplinkNetwork:      getUplinkNetworkSpecs(node),
+	}
+
+	mgmtNetworkRaw := node["management_network"].([]interface{})
+	if len(mgmtNetworkRaw) > 0 {
+		mgmtNetworkData := mgmtNetworkRaw[0].(map[string]interface{})
+		name := mgmtNetworkData["portgroup_name"].(string)
+		vlan := int32(mgmtNetworkData["vlan_id"].(int))
+
+		nodeSpec.VMManagementPortgroupName = &name
+		nodeSpec.VMManagementPortgroupVlan = &vlan
 	}
 
 	return nodeSpec

--- a/internal/nsx_edge_cluster/edge_node_subresource.go
+++ b/internal/nsx_edge_cluster/edge_node_subresource.go
@@ -84,6 +84,28 @@ func EdgeNodeSchema() *schema.Resource {
 				Description:  "The gateway address for the management network",
 				ValidateFunc: validationUtils.ValidateIPv4AddressSchema,
 			},
+			"management_network": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: "The management network which will be created for this node",
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"portgroup_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Description:  "The name of the portgroup",
+							ValidateFunc: validation.NoZeroValues,
+						},
+						"vlan_id": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							Description:  "The VLAN ID for the portgroup",
+							ValidateFunc: validation.IntBetween(0, 4095),
+						},
+					},
+				},
+			},
 			"first_nsx_vds_uplink": {
 				Type:         schema.TypeString,
 				Optional:     true,

--- a/internal/provider/resource_edge_cluster.go
+++ b/internal/provider/resource_edge_cluster.go
@@ -65,13 +65,13 @@ func ResourceEdgeCluster() *schema.Resource {
 			},
 			"tier0_name": {
 				Type:         schema.TypeString,
-				Required:     true,
+				Optional:     true,
 				Description:  "Name for the Tier-0 gateway",
 				ValidateFunc: validation.NoZeroValues,
 			},
 			"tier1_name": {
 				Type:         schema.TypeString,
-				Required:     true,
+				Optional:     true,
 				Description:  "Name for the Tier-1 gateway",
 				ValidateFunc: validation.NoZeroValues,
 			},
@@ -90,7 +90,8 @@ func ResourceEdgeCluster() *schema.Resource {
 			},
 			"routing_type": {
 				Type:         schema.TypeString,
-				Required:     true,
+				Optional:     true,
+				RequiredWith: []string{"tier0_name"},
 				Description:  "One among: EBGP, STATIC",
 				ValidateFunc: validation.StringInSlice([]string{"EBGP", "STATIC"}, false),
 			},
@@ -102,7 +103,8 @@ func ResourceEdgeCluster() *schema.Resource {
 			},
 			"high_availability": {
 				Type:         schema.TypeString,
-				Required:     true,
+				Optional:     true,
+				RequiredWith: []string{"tier0_name"},
 				Description:  "One among: ACTIVE_ACTIVE, ACTIVE_STANDBY",
 				ValidateFunc: validation.StringInSlice([]string{"ACTIVE_ACTIVE", "ACTIVE_STANDBY"}, false),
 			},

--- a/internal/provider/resource_edge_cluster_test.go
+++ b/internal/provider/resource_edge_cluster_test.go
@@ -109,7 +109,7 @@ func getEdgeClusterConfigExpansion() string {
 		"192.168.19.3/24")
 	edgeNode3 := getEdgeNodeConfig(
 		edgeNode3Name,
-		"10.0.0.55/24",
+		"10.0.0.54/24",
 		"192.168.52.16/24",
 		"192.168.52.17/24",
 		"192.168.18.6/24",

--- a/internal/provider/resource_network_pool.go
+++ b/internal/provider/resource_network_pool.go
@@ -69,7 +69,7 @@ func ResourceNetworkPool() *schema.Resource {
 						"vlan_id": {
 							Type:        schema.TypeInt,
 							Description: "VLAN ID associated with the network",
-							Optional:    true,
+							Required:    true,
 						},
 						"ip_pools": {
 							Type:        schema.TypeList,

